### ISSUE TITLE
existential impl trait overlooks "hidden" lifetimes from projections

### ISSUE THREAD

**Issue by nikomatsakis** [rust-lang/rust#51525]
In this example ([play](https://play.rust-lang.org/?gist=b02b15c68b0f6576fdf22b81ae7421b3&version=nightly&mode=debug)), we currently get an incorrect compilation error:

```rust
trait Foo<'a> {
    type Bar;
}

impl<'a> Foo<'a> for u32 {
    type Bar = &'a ();
}

fn baz<'a, T>() -> impl IntoIterator<Item = T::Bar>
where T: Foo<'a> { 
    None
}

fn main() { }
```

yields

```
error[E0308]: mismatched types
  --> src/main.rs:11:5
   |
11 |     None
   |     ^^^^ lifetime mismatch
   |
   = note: expected type `Foo<'static>`
              found type `Foo<'a>`
```

This does not seem correct. =)  The problem is that `T::Bar` (at HIR lowering time) does not mention `'a`. So we wind up mapping this -- because of the way the region desugaring works -- to `<T as Foo<'static>>::Bar`. If you manually write `<T as Foo<'a>>::Bar` in the original Rust source, the `'a` becomes visible and is lowered correctly. In other words, [this version compiles](https://play.rust-lang.org/?gist=1a1423e1b69fcc8bf1b0cd693a5e207e&version=nightly&mode=debug):

```rust
trait Foo<'a> {
    type Bar;
}

impl<'a> Foo<'a> for u32 {
    type Bar = &'a ();
}

fn baz<'a, T>() -> impl IntoIterator<Item = <T as Foo<'a>>::Bar>
where T: Foo<'a> { 
    None
}

fn main() { }
```

**Comment by nikomatsakis** [comment]
cc @cramertj @oli-obk

**Comment by Dylan-DPC** [comment]
This code now compiles, so requires a test to close this issue
